rootston/seat.c

```diff
@@ -518,9 +691,153 @@ static void seat_add_touch(struct roots_seat *seat,
+	struct roots_seat *seat = tablet_pad->seat;
+
+	wl_list_remove(&tablet_pad->device_destroy.link);
+	wl_list_remove(&tablet_pad->tablet_destroy.link);
+	wl_list_remove(&tablet_pad->attach.link);
+	wl_list_remove(&tablet_pad->link);
+	wl_list_remove(&tablet_pad->tablet_link);
+
+	wl_list_remove(&tablet_pad->button.link);
+	wl_list_remove(&tablet_pad->strip.link);
+	wl_list_remove(&tablet_pad->ring.link);
+	free(tablet_pad);
+
+	seat_update_capabilities(seat);
+}
+
+static void handle_pad_tool_destroy(struct wl_listener *listener, void *data) {
+	struct roots_tablet_pad *pad =
+		wl_container_of(listener, pad, tablet_destroy);
+
+	pad->tablet = NULL;
+	wl_list_remove(&pad->tablet_link);
+	wl_list_init(&pad->tablet_link);
+
+	wl_list_remove(&pad->tablet_destroy.link);
+	wl_list_init(&pad->tablet_destroy.link);
+}
+
+static void attach_tablet_pad(struct roots_tablet_pad *pad,
+		struct roots_tablet_tool *tool) {
+	wlr_log(L_DEBUG, "Attaching tablet pad \"%s\" to tablet tool \"%s\"",
+		pad->device->name, tool->device->name);
+
+	pad->tablet = tool;
+	wl_list_remove(&pad->tablet_link);
+	wl_list_insert(&tool->pads, &pad->tablet_link);
+
+	wl_signal_add(&tool->device->events.destroy,
+		&pad->tablet_destroy);
+}
+
+static void handle_tablet_pad_attach(struct wl_listener *listener, void *data) {
+	struct roots_tablet_pad *pad =
+		wl_container_of(listener, pad, attach);
+	struct wlr_tablet_tool *wlr_tool = data;
+	struct roots_tablet_tool *tool = wlr_tool->data;
+
+	attach_tablet_pad(pad, tool);
+}
+
+static void handle_tablet_pad_ring(struct wl_listener *listener, void *data) {
+	struct roots_tablet_pad *pad =
+		wl_container_of(listener, pad, ring);
+	struct wlr_event_tablet_pad_ring *event = data;
+
+	wlr_send_tablet_v2_tablet_pad_ring(pad->tablet_v2_pad,
+		event->ring, event->position,
+		event->source == WLR_TABLET_PAD_RING_SOURCE_FINGER,
+		event->time_msec);
+}
+
+static void handle_tablet_pad_strip(struct wl_listener *listener, void *data) {
+	struct roots_tablet_pad *pad =
+		wl_container_of(listener, pad, strip);
+	struct wlr_event_tablet_pad_strip *event = data;
+
+	wlr_send_tablet_v2_tablet_pad_strip(pad->tablet_v2_pad,
+		event->strip, event->position,
+		event->source == WLR_TABLET_PAD_STRIP_SOURCE_FINGER,
+		event->time_msec);
+}
+
+static void handle_tablet_pad_button(struct wl_listener *listener, void *data) {
+	struct roots_tablet_pad *pad =
+		wl_container_of(listener, pad, button);
+	struct wlr_event_tablet_pad_button *event = data;
+
+	wlr_send_tablet_v2_tablet_pad_mode(pad->tablet_v2_pad,
+		event->group, event->mode, event->time_msec);
+
+	wlr_send_tablet_v2_tablet_pad_button(pad->tablet_v2_pad,
+		event->button, event->time_msec,
+		(enum zwp_tablet_pad_v2_button_state)event->state);
+}
+
 static void seat_add_tablet_pad(struct roots_seat *seat,
 		struct wlr_input_device *device) {
-	// TODO
+	struct roots_tablet_pad *tablet_pad =
+		calloc(sizeof(struct roots_tablet_pad), 1);
```

Comment:
Nitpick: doesn't matter a lot, but strictly speaking, the order of the arguments are swapped.